lib/palletjack/tool.rb

```diff
@@ -342,29 +342,42 @@ def pallet_dir(kind, name)
       config_dir :warehouse, kind, name
     end
 
-    # Write a key box file inside a pallet
+    # :call-seq:
+    # pallet_box kind, name, box, 'key.path' => value, ...
+    # pallet_box kind, name, box { { key: { path: value, ... } } }
+    #
+    # Write keys to a box file inside a pallet
+    #
+    # Any key.path assignments given as parameters, and the hash value
+    # returned from a block will be merged into the named box file.
     #
-    # The block should return a hash representing the contents of the box.
     # All keys will be stringified, so we can use key: short forms for
     # declaration of the box contents.
     #
     # Uses config_file to create the file, so any symbols will
     # be looked up in the options hash.
     #
-    # N.B! If the box already exists, it will be overwritten!
-    #--
-    # FIXME: should new values be merged with existing box data instead?
-    #++
-    #
     # Example:
     #
     #   pallet_box 'domain', :domain, 'dns' do
     #     { dns:{ ns:options[:soa_ns].split(',') } }
     #   end
+    #--
+    # FIXME: Box I/O should probably be managed by PalletJack::Pallet,
+    #        but that requires some redesign work.
+    #++
+
+    def pallet_box(kind, name, box, keys = {}, &block)
+      box = options[box] if box.is_a?(Symbol)
+      box_path = config_path(:warehouse, kind, name, "#{box}.yaml")
+      contents = KVDAG::KeyPathHashProxy.new
+
+      contents.merge! YAML::load_file(box_path) if box_path.file?
+      keys.each { |key, value| contents[key] = value }
+      contents.merge! block.call if block_given?
 
-    def pallet_box(kind, name, box, &block)
-      config_file :warehouse, kind, name, "#{box}.yaml" do |box_file|
-        box_file << block.call.deep_stringify_keys.to_yaml
+      config_file box_path do |box_file|
+        box_file << contents.deep_stringify_keys.to_yaml
```

Comment:
Do we want to care about doing overwrites properly? When I wanted to do so the other day, I borrowed some code from the standard library (since it of course wasn't structured for any old user code to use): creideiki/post_images_bot@793d66f6cb64ce25636e693fc7bdba4a369ab5cc